File: keno.py
from flask import Flask, request
import requests
import random
import re
from time import sleep

app = Flask(__name__)

def parse_keno_numbers(s):
    # Use a regular expression to find all sequences of digits in the string
    numbers = re.findall(r'\d+', s)
    valid_numbers = []

    for number in numbers:
        # Convert string to integer
        num = int(number)

        # Check if number is a valid keno number
        if 1 <= num <= 80:
            valid_numbers.append(num)

            # Skip the length of this number in the loop to avoid overlapping
            # (e.g., '54' in '154' should not allow '4' to be extracted again)
            continue

    return valid_numbers


def fetch_seed():
    try:
        # hardware random generator 1
        response = requests.get("https://rand.kroy.io")
        response.raise_for_status()
        return response.text
    except requests.exceptions.RequestException as e:
        try:
            # hardware random generator 2
            random_org_response = requests.get("https://www.random.org/integers/?num=1&min=1&max=1000000000&col=1&base=10&format=plain&rnd=new")
            random_org_response.raise_for_status()
            return random_org_response.text
        except requests.exceptions.RequestException as e:
            return str(random.randint(1, 1000000000))

def generate_unique_random_numbers(seed, count):
    random.seed(seed)
    return random.sample(range(1, 81), count)

# Flip a coin `times` times.  Make it hacky so it's quick
def flip_coin(times):
    results = [random.getrandbits(1) for _ in range(times)]
    heads_count = sum(results)
    # Heads == true, tails == false
    return heads_count > (times - heads_count)

@app.route('/', methods=['GET'])
def random_numbers():
    count = request.args.get('count', default=4, type=int)
    count = max(1, min(count, 20))
    seed = fetch_seed()
    numbers = generate_unique_random_numbers(seed, count)

    sleep(count * .01)
    seed = fetch_seed()
    threespot = generate_unique_random_numbers(seed, 3)

    sleep(count * .001)
    seed = fetch_seed()
    alternate = generate_unique_random_numbers(seed, count)
    
    heads_wins = flip_coin(574673)
    coin_flip = "A" if heads_wins else "B" 
    
    # dark mode because fuckoff
    html = '''
    <html>
    <head>
        <style>
            body {{
                background-color: #333;
                color: #fff;
                font-family: Arial, sans-serif;
                margin: 40px;
            }}
            hr {{
                border: 1px solid #555;
            }}
        </style>
    </head>
    <body>
        <h2>Keno Picker</h2>
        Seed: {seed}<br><hr><br>Result A:<br> {number_list}
        <br><hr><br>Three Spot:<br> {threespot_list}
        <br><hr><br>Result B:<br> {alternate_list}
        <br><hr><br>Which result to pick: {result}
    </body>
    </html>
    '''.format(seed=seed, alternate_list='<br>'.join(map(str, alternate)), threespot_list='<br>'.join(map(str, threespot)), number_list='<br>'.join(map(str, numbers)), result=coin_flip)

    
    return html

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000)

File: keno.go
package main

import (
	"crypto/rand"
	"encoding/binary"
	"fmt"
	"io"
	"log"
	"math/big"
	mrand "math/rand"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
)

func fetchSeed() (int64, string) {
	// Attempt to fetch the seed from https://rand.kroy.io/
	seed, err := fetchSeedFromURL("https://rand.kroy.io/")
	if err == nil {
		return seed, "rand.kroy.io"
	}

	// If that fails, attempt to fetch the seed from random.org
	seed, err = fetchSeedFromURL("https://www.random.org/integers/?num=1&min=1&max=1000000&col=1&base=10&format=plain&rnd=new")
	if err == nil {
		return seed, "random.org"
	}

	// If both external sources fail, fall back to on-device generation
	return generateOnDeviceSeed(), "ondevice"
}

func fetchSeedFromURL(url string) (int64, error) {
	client := http.Client{
		Timeout: 4 * time.Second,
	}

	resp, err := client.Get(url)
	if err != nil {
		fmt.Println("Failed to fetch seed from URL:", url, err)
		return 0, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("Failed to read response body:", err)
		return 0, err
	}

	seedStr := strings.TrimSpace(string(body))
	seed, err := strconv.ParseInt(seedStr, 10, 64)
	if err != nil {
		fmt.Println("Failed to parse seed:", err)
		return 0, err
	}

	return seed, nil
}

func generateOnDeviceSeed() int64 {
	// Create a byte slice to hold the random bytes
	seedBytes := make([]byte, 8) // 8 bytes = 64 bits

	// Read random bytes from crypto/rand
	_, err := rand.Read(seedBytes)
	if err != nil {
		fmt.Println("Failed to generate random bytes:", err)
		return 0
	}

	// Convert the random bytes to an int64
	seed := int64(binary.LittleEndian.Uint64(seedBytes))
	return seed
}

func generateUniqueRandomNumbers(rng *mrand.Rand, count int) []int {
	numbers := make([]int, 0, count)
	seen := make(map[int]bool)
	for len(numbers) < count {
		num := rng.Intn(80) + 1 // Generate numbers between 1 and 80
		if !seen[num] {
			seen[num] = true
			numbers = append(numbers, num)
		}
	}
	return numbers
}

func flipCoin(rng *mrand.Rand, times int) bool {
	headsCount := 0
	for i := 0; i < times; i++ {
		if rng.Intn(2) == 0 { // 0 for heads, 1 for tails
			headsCount++
		}
	}
	return headsCount > times-headsCount
}

func main() {
	r := gin.Default()

	r.LoadHTMLGlob("/app/templates/*")

	r.GET("/", func(c *gin.Context) {
		countStr := c.DefaultQuery("count", "4")
		count, err := strconv.Atoi(countStr)
		if err != nil || count < 1 || count > 20 {
			count = 4 // Default to 4 if any error occurs
		}

		jsonOutput := c.DefaultQuery("json", "0") // Check for json

		seed, source := fetchSeed()
		fmt.Println("seed: ", seed)

		// Create a new RNG instance using the fetched seed
		rng := mrand.New(mrand.NewSource(seed))

		numbers := generateUniqueRandomNumbers(rng, count)
		threespot := generateUniqueRandomNumbers(rng, 3)
		sixspot := generateUniqueRandomNumbers(rng, 6)
		alternate := generateUniqueRandomNumbers(rng, count)

		// Generate a random number of flips between 1 and 1,000,000 using crypto/rand
		maxFlips := big.NewInt(1000000)
		n, err := rand.Int(rand.Reader, maxFlips)
		if err != nil {
			log.Fatalf("Failed to generate random number of flips: %v", err)
		}

		two := big.NewInt(2)
		remainder := new(big.Int).Mod(n, two)
		flipped := remainder.Cmp(big.NewInt(0)) == 0

		if jsonOutput == "1" {
			c.JSON(http.StatusOK, gin.H{
				"seed":      seed,
				"numbers":   numbers,
				"threespot": threespot,
				"sixspot":   sixspot,
				"alternate": alternate,
				"coinFlip":  flipped,
				"flipped":   n,
				"source":    source,
			})
		} else {
			c.HTML(http.StatusOK, "index.tmpl", gin.H{
				"seed":      seed,
				"numbers":   numbers,
				"alternate": alternate,
				"threespot": threespot,
				"sixspot":   sixspot,
				"coinFlip":  flipped,
				"flipped":   n,
				"source":    source,
			})
		}
	})

	r.Run(":5000") // Listen and serve on 0.0.0.0:5000
}
